raiden/utils/nursery.py

```diff
@@ -139,20 +147,29 @@ def subprocess_stopped(result: AsyncResult) -> None:
             def spawn_under_watch(function: Callable, *args: Any, **kwargs: Any) -> Greenlet:
                 greenlet = gevent.spawn(function, *args, **kwargs)
 
-                def kill_grenlet(_result: AsyncResult) -> None:
-                    # gevent.GreenletExit must **not** be used here, since that
-                    # exception is considered a successful run and it is not
-                    # re-raised on calls to `get`
-                    exception = RuntimeError("Janitor is stopping")
-                    greenlet.throw(exception)
-
-                # The Event.rawlink is executed inside the Hub thread, which
-                # does validation and *raises on blocking calls*, to go around
-                # this a new greenlet has to be spawned, that in turn will
-                # raise the exception.
-                callback = SpawnedLink(kill_grenlet)
+                # The callback provided to `AsyncResult.rawlink` is executed
+                # inside the Hub thread, the callback is calling `throw` which
+                # has to be called from the Hub, so here there is no need to
+                # wrap the callback in a SpawnedLink.
+                #
+                # `throw` does not raise the exception if the greenlet has
+                # finished, which is exactly the semantics needed here.
+                def stop_greenlet_from_hub(result: AsyncResult) -> None:
```

Comment:
Note that this code could also be moved to the `_free_resources` function, as to make cleanup more local, however I did not do this because I'm no sure if the Hub would be running when `atexit` handlers are executed. It is actually an interesting question if the current code works as expected since the Popen class is monkey patched.